eng/common/scripts/Update-DocsMsToc.ps1

```diff
@@ -0,0 +1,192 @@
+
+# This criteria is different from criteria used in `Update-DocsMsPackages.ps1`
+# because we need to generate ToCs for packages which are not necessarily "New"
+# in the metadata AND onboard legacy packages (which `Update-DocsMsPackages.ps1`
+# does not do)
+$metadata = (Get-CSVMetadata).Where({
+    $_.Package `
+      -and $onboardedPackages.ContainsKey($_.Package) `
+      -and $_.Hide -ne 'true'
+  })
+
+# Map metadata to package. Validate metadata entries.
+$packagesForToc = @{}
+foreach ($packageName in $onboardedPackages.Keys) {
+  $metadataEntry = $metadata.Where({ $_.Package -eq $packageName })
+
+  if (!$metadataEntry) {
+    LogWarning "Could not find metadata for package $packageName. Skipping"
+    continue
+  }
+  if ($metadataEntry -is [System.Collections.IEnumerable] -and ($metadataEntry | Measure-Object).Count -gt 1) {
+    LogWarning "Duplicate metadata for package $packageName. Using the entry which appears first in metadata CSV file."
+  }
+  $metadataEntry = $metadataEntry[0]
+
+  if (!$metadataEntry.ServiceName) {
+    LogWarning "Empty ServiceName for package $packageName. Skipping."
+    continue
+  }
+
+  $packagesForToc[$packageName] = $metadataEntry
+}
+
+# Get unique service names and sort alphabetically to act as the service nodes
+# in the ToC
+$services = @{}
+foreach ($package in $packagesForToc.Values) {
+  if ($package.ServiceName -eq 'Core') {
+    # Skip packages under the service category "Core". Those will be handled
+    # later
+    continue
+  }
+  if (!$services.ContainsKey($package.ServiceName)) {
+    $services[$package.ServiceName] = $true
+  }
+}
+$serviceNameList = $services.Keys | Sort-Object
+
+
+$toc = @()
+foreach ($service in $serviceNameList) {
+  Write-Host "Building service: $service"
+
+  $packageItems = @()
+  $packageCount = 0
+
+  # Client packages get individual entries
+  $clientPackages = $packagesForToc.Values.Where({ $_.ServiceName -eq $service -and (@('client', '') -contains $_.Type) })
+  $clientPackages = $clientPackages | Sort-Object -Property Package
+  if ($clientPackages) {
+    foreach ($clientPackage in $clientPackages) {
+      $packageItems += GetClientPackageNode -clientPackage $clientPackage
+      $packageCount += 1
+    }
+  }
+
+
+  # All management packages go under a single `Management` header in the ToC
+  $mgmtPackages = $packagesForToc.Values.Where({ $_.ServiceName -eq $service -and ('mgmt' -eq $_.Type) })
+  $mgmtPackages = $mgmtPackages | Sort-Object -Property Package
+  if ($mgmtPackages) {
+    $children = &$GetDocsMsTocChildrenForManagementPackages `
+      -packageMetadata $mgmtPackages `
+      -docRepoLocation $DocRepoLocation
+
+    $packageItems += [PSCustomObject]@{
+      name     = 'Management'
+      # There could be multiple packages, ensure this is treated as an array
+      # even if it is a single package
+      children = @($children)
+    };
+    $packageCount += 1
+  }
+
+  $serviceReadmeBaseName = $service.ToLower().Replace(' ', '-')
```

Comment:
Does this logic need to be insync with the place we produce the service md? Are they basically the same?

Reply:
The logic should be the same however we don't have automation which produces service-level markdown files at this time.